state/watcher.go

```diff
@@ -3218,14 +3218,152 @@ func ensureSuffixFn(marker string) func(string) string {
 	}
 }
 
+// watchActionNotificationsFilteredBy starts and returns a StringsWatcher
+// that notifies on new Actions being enqueued on the ActionRecevers
+// being watched as well as changes to non-completed Actions.
+func (st *State) watchActionNotificationsFilteredBy(receivers ...ActionReceiver) StringsWatcher {
+	return newActionNotificationWatcher(st, false, receivers...)
+}
+
 // watchEnqueuedActionsFilteredBy starts and returns a StringsWatcher
 // that notifies on new Actions being enqueued on the ActionRecevers
 // being watched.
 func (st *State) watchEnqueuedActionsFilteredBy(receivers ...ActionReceiver) StringsWatcher {
-	return newCollectionWatcher(st, colWCfg{
-		col:    actionNotificationsC,
-		filter: makeIdFilter(st, actionMarker, receivers...),
-		idconv: actionNotificationIdToActionId,
+	return newActionNotificationWatcher(st, true, receivers...)
+}
+
+// actionNotificationWatcher is a StringsWatcher that watches for changes on the
+// action notification collection, but only triggers events once per action.
+type actionNotificationWatcher struct {
+	commonWatcher
+	source   chan watcher.Change
+	sink     chan []string
+	filter   func(interface{}) bool
+	fireOnce bool
```

Comment:
to save the reader groking the code, the you add a doc line to explain what this is for